### PR TITLE
Fix EventPipe IPC streaming

### PR DESCRIPTION
**Pull request by sywhang**
This addresses https://github.com/dotnet/coreclr/issues/23756.

In particular, it decreases`EventPipeFile`'s `m_block`'s size from 100K to 1K for IPC streaming sessions *only*.

### CODE REVIEW

File: src/vm/fastserializer.h [hunk 1]
```diff
@@ -36,6 +36,12 @@ enum class FastSerializerTags : BYTE
     Limit                   // Just past the last valid tag, used for asserts.
 };
 
+enum class StreamWriterType : BYTE
```
Comment (jorive) [inline src/vm/fastserializer.h]: nit: BYTE -> uint8_t (Prefer C++ types over macros)

File: src/vm/fastserializer.h [hunk 2]
```diff
@@ -45,6 +51,10 @@ class StreamWriter
     StreamWriter() = default;
     virtual ~StreamWriter() = default;
     virtual bool Write(const void *lpBuffer, const uint32_t nBytesToWrite, uint32_t &nBytesWritten) const = 0;
+    StreamWriterType GetType() { return m_type; }
+
+protected:
+    StreamWriterType m_type;
```
Comment (jorive) [inline src/vm/fastserializer.h]: Please change field to `const StreamWriterType m_type`, then:
- ctor becomes:
```cpp
StreamWriter(StreamWriterType type) : m_type(type)
FileStreamWriter::FileStreamWriter(const SString &outputFilePath) : StreamWriter(StreamWriterType::File)
IpcStreamWriter::IpcStreamWriter(IpcStream *pStream) : StreamWriter(StreamWriterType::Ipc), _pStream(pStream)
```

File: src/vm/eventpipefile.cpp [hunk 3]
```diff
@@ -19,7 +19,14 @@ EventPipeFile::EventPipeFile(StreamWriter *pStreamWriter) : FastSerializableObje
     }
     CONTRACTL_END;
 
-    m_pBlock = new EventPipeBlock(100 * 1024);
+    if (pStreamWriter->GetType() == StreamWriterType::Ipc)
```
Comment (jorive) [inline src/vm/eventpipefile.cpp]:
```cpp
m_pBlock = new EventPipeBlock(
    pStreamWriter->GetType() == StreamWriterType::Ipc ? 1024 : 1024 * 1024);
```

File: src/vm/eventpipebuffermanager.h [hunk 4]
```diff
@@ -105,6 +105,10 @@ class EventPipeBufferManager
     // skip any events that might be partially written due to races when tracing is stopped.
     void WriteAllBuffersToFile(EventPipeFile *pFastSerializableObject, LARGE_INTEGER stopTimeStamp);
 
+    // Used by IPC streaming EventPipe sessions. Writes out a single event out to 
+    // the IPC channel. 
+    void EventPipeBufferManager::WriteNextEventToFile(EventPipeFile *pFile, LARGE_INTEGER stopTimeStamp);
```
Comment (jorive) [inline src/vm/eventpipebuffermanager.h]: Do not need `EventPipeBufferManager::` on the declaration.

File: src/vm/eventpipefile.cpp (hunk 3 above)
Comment (jorive) [reply src/vm/eventpipefile.cpp]: Could you please move the conditional to the actual `maxBlockSize` data?

File: src/vm/eventpipe.cpp [hunk 5]
```diff
@@ -615,7 +615,7 @@ void WINAPI EventPipe::FlushTimer(PVOID parameter, BOOLEAN timerFired)
         // the current timestamp are written into the file.
         LARGE_INTEGER stopTimeStamp;
         QueryPerformanceCounter(&stopTimeStamp);
-        s_pBufferManager->WriteAllBuffersToFile(s_pFile, stopTimeStamp);
+        s_pBufferManager->WriteNextEventToFile(s_pFile, stopTimeStamp);
```
Comment (noahfalk) [inline src/vm/eventpipe.cpp]: Have you tested this on a scenario that produces many events? I think this change limited the IPC channel to sending at most 20 events/sec : ) I recommend that you send all currently buffered events, including partially filled buffers.

File: src/vm/eventpipefile.cpp [hunk 6]
```diff
@@ -19,7 +19,7 @@ EventPipeFile::EventPipeFile(StreamWriter *pStreamWriter) : FastSerializableObje
     }
     CONTRACTL_END;
 
-    m_pBlock = new EventPipeBlock(100 * 1024);
+    m_pBlock = (pStreamWriter->GetType() == StreamWriterType::Ipc ? new EventPipeBlock(1024) : new EventPipeBlock(1024 * 100));
```
Comment (noahfalk) [inline src/vm/eventpipefile.cpp]: I looked into this a little more, and I am suspicious this is going to cause you some problems too : ) There is an assumption in the code that an EventBlock is larger than the biggest event and events include stacks. Assuming 8 bytes per stack IP all you would need is a 128 frame stack to exceed the 1KB block size and then you would get an error. At the same time if you wrote very small events, say 20 bytes, once per second you would still have to wait ~50 seconds before they got flushed.

I think what can solve this is to add an API to EventPipeFile that flushes the current block regardless of how full it is. EventPipeFile::WriteEnd() is very close to that, except it also writes the terminator that ends the file.  Once you have that API, update WriteBuffersToFile() to call the new file flush API at the end of its work.

**Pull request merged**